src/App.tsx

```diff
@@ -15,48 +14,93 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  Default = 'default',
+  Alphabetically = 'alphabetically',
+  Length = 'length',
+  Reverse = 'reverse',
+}
+
 export const App: React.FC = () => {
+  const [copyArray, setArray] = useState([...goodsFromServer]);
+  const [currentSort, setCurrentSort] = useState<SortType>(SortType.Default);
+  const [isReversed, setIsReversed] = useState(false);
+
+  const isOriginalOrder = () => {
+    return JSON.stringify(copyArray) === JSON.stringify(goodsFromServer);
+  };
+
+  const sortAlphabetically = () => {
+    const sortedArray = [...goodsFromServer].sort((a, b) => a.localeCompare(b));
+
+    setArray(isReversed ? sortedArray.reverse() : sortedArray);
+    setCurrentSort(SortType.Alphabetically);
+  };
+
+  const sortByLength = () => {
+    const sortedArray = [...goodsFromServer].sort(
+      (a, b) => a.length - b.length,
+    );
+
+    setArray(isReversed ? sortedArray.reverse() : sortedArray);
+    setCurrentSort(SortType.Length);
+  };
+
+  const reverse = () => {
+    setArray([...copyArray].reverse());
+    setIsReversed(prev => !prev);
+  };
+
+  const resetButton = () => {
+    setArray([...goodsFromServer]);
+    setCurrentSort(SortType.Default);
+    setIsReversed(false);
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
+          onClick={sortAlphabetically}
           type="button"
-          className="button is-info is-light"
+          className={`button is-info ${currentSort === SortType.Alphabetically ? '' : 'is-light'}`}
         >
           Sort alphabetically
         </button>
 
         <button
+          onClick={sortByLength}
           type="button"
-          className="button is-success is-light"
+          className={`button is-info ${currentSort === SortType.Length ? '' : 'is-light'}`}
         >
           Sort by length
         </button>
 
         <button
+          onClick={reverse}
           type="button"
-          className="button is-warning is-light"
+          className={`button is-info ${isReversed ? '' : 'is-light'}`}
         >
           Reverse
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {!isOriginalOrder() && (
+          <button
+            type="button"
+            className="button is-danger is-light"
+            onClick={resetButton}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {copyArray.map((item, index) => (
+          <li key={index} data-cy="Good">
```

Comment:
Using the index as a key in a list can lead to performance issues and problems with component identity. It's better to use a unique identifier for each item if possible.